demos/bluetooth/mesh/thingy52/thingy52_music_cli/src/model_handler.c

```diff
@@ -0,0 +1,355 @@
+	[0 ... NUM_FREQ_BANDS - 1] = BT_MESH_THINGY52_CLI_INIT
+};
+
+static inline int nrfx_err_code_check(nrfx_err_t nrfx_err)
+{
+	return NRFX_ERROR_BASE_NUM - nrfx_err ? true : false;
+}
+
+static void button_handler(uint32_t pressed, uint32_t changed)
+{
+	if (pressed & BIT(0)) {
+		int err = 0;
+		static bool onoff = true;
+
+		if (onoff) {
+			err |= gpio_port_set_bits_raw(io_expander,
+						      BIT(MIC_PWR));
+			err |= nrfx_err_code_check(nrfx_pdm_start());
+
+			for (int i = GREEN_LED; i <= RED_LED; i++) {
+				err |= sx1509b_pwm_pin_set(io_expander, i, 20);
+			}
+
+			k_delayed_work_submit(&mic_data.microphone_work,
+					      K_NO_WAIT);
+		} else {
+			k_delayed_work_cancel(&mic_data.microphone_work);
+
+			for (int i = GREEN_LED; i <= RED_LED; i++) {
+				err |= sx1509b_pwm_pin_set(io_expander, i, 0);
+			}
+
+			err |= nrfx_err_code_check(nrfx_pdm_stop());
+			err |= gpio_port_clear_bits_raw(io_expander,
+							BIT(MIC_PWR));
+		}
+
+		onoff = !onoff;
+
+		if (err) {
+			printk("Error running PDM\n");
+		}
+	}
+}
+
+static void button_and_led_init(void)
+{
+	int err = 0;
+
+	err |= dk_buttons_init(button_handler);
+
+	for (int i = GREEN_LED; i <= RED_LED; i++) {
+		err |= sx1509b_pwm_pin_configure(io_expander, i);
+	}
+
+	if (err) {
+		printk("Initializing buttons and leds failed\n");
+	}
+}
+
+static void pdm_event_handler(nrfx_pdm_evt_t const *p_evt)
+{
+	if (p_evt->error) {
+		printk("PDM overflow error\n");
+	} else if (p_evt->buffer_requested && p_evt->buffer_released == 0) {
+		nrfx_pdm_buffer_set(mic_data.pdm_buf_a,
+				    sizeof(mic_data.pdm_buf_a) /
+					    sizeof(mic_data.pdm_buf_a[0]));
+	} else if (p_evt->buffer_requested &&
+		   p_evt->buffer_released == mic_data.pdm_buf_a) {
+		nrfx_pdm_buffer_set(mic_data.pdm_buf_b,
+				    sizeof(mic_data.pdm_buf_b) /
+					    sizeof(mic_data.pdm_buf_b[0]));
+
+		if (!mic_data.fft_result_ready) {
+			arm_rfft_q15(&S15, (q15_t *)mic_data.pdm_buf_a,
+				     mic_data.fft_buf);
+			arm_cmplx_mag_q15(mic_data.fft_buf, mic_data.spectrum,
+					  SPECTRUM_SIZE);
+
+			mic_data.fft_result_ready = true;
+		}
+	} else if (p_evt->buffer_requested &&
+		   p_evt->buffer_released == mic_data.pdm_buf_b) {
+		nrfx_pdm_buffer_set(mic_data.pdm_buf_a,
+				    sizeof(mic_data.pdm_buf_a) /
+					    sizeof(mic_data.pdm_buf_a[0]));
+
+		if (!mic_data.fft_result_ready) {
+			arm_rfft_q15(&S15, (q15_t *)mic_data.pdm_buf_b,
+				     mic_data.fft_buf);
+			arm_cmplx_mag_q15(mic_data.fft_buf, mic_data.spectrum,
+					  SPECTRUM_SIZE);
+
+			mic_data.fft_result_ready = true;
+		}
+	}
+}
+
+static void send_color_helper(int iterator, uint16_t value)
+{
+	int err = 0;
+
+	static struct bt_mesh_thingy52_rgb_msg msg = {
+		.ttl = 0,
+		.delay = 0xFFFF,
+		.speaker_on = false,
+	};
+
+	/* Should match NUM_FREQ_BANDS */
+	if (iterator == 0) {
+		msg.color.red = 0;
+		msg.color.green = 0;
+		msg.color.blue = value;
+		err |= bt_mesh_thingy52_cli_rgb_set(&thingy52_cli[iterator],
+						    NULL, &msg);
```

Comment:
This `err |=` pattern is generally frowned upon, as it corrupts the error and allows the program flow to continue even if things went wrong. I'm not gonna make you change it everywhere, but here it's not even needed.